Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,175 @@
+
+    private var state: ClientState
+    
+    var proxyEstablished: Bool {
+        switch self.state {
+        case .active:
+            return true
+        case .error, .inactive, .waitingForAuthenticationMethod, .waitingForClientGreeting, .waitingForClientRequest, .waitingForServerResponse:
+            return false
+        }
+    }
+    
+    var shouldBeginHandshake: Bool  {
+        switch self.state {
+        case .inactive:
+            return true
+        case .active, .error, .waitingForAuthenticationMethod, .waitingForClientGreeting, .waitingForClientRequest, .waitingForServerResponse:
+            return false
+        }
+    }
+    
+    init() {
+        self.state = .inactive
+    }
+    
+}
+
+// MARK: - Incoming
+extension ClientStateMachine {
+    
+    mutating func receiveBuffer(_ buffer: inout ByteBuffer) throws -> ClientAction {
+        do {
+            switch self.state {
+            case .waitingForAuthenticationMethod(let greeting):
+                return try self.handleSelectedAuthenticationMethod(&buffer, greeting: greeting)
+            case .waitingForServerResponse(let request):
+                return try self.handleServerResponse(&buffer, request: request)
+            case .active, .error, .inactive, .waitingForClientGreeting, .waitingForClientRequest:
+                throw SOCKSError.UnexpectedRead()
+            }
+        } catch {
+            self.state = .error
+            throw error
+        }
+    }
+    
+    mutating func handleSelectedAuthenticationMethod(_ buffer: inout ByteBuffer, greeting: ClientGreeting) throws -> ClientAction {
+        return try buffer.parseUnwindingIfNeeded { buffer -> ClientAction in
+            guard let selected = try buffer.readMethodSelection() else {
+                return .waitForMoreData
+            }
+            guard greeting.methods.contains(selected.method) else {
+                throw SOCKSError.InvalidAuthenticationSelection(selection: selected.method)
+            }
+                
+            // we don't current support any form of authentication
+            return self.authenticate(&buffer)
+        }
+    }
+    
+    mutating func handleServerResponse(_ buffer: inout ByteBuffer, request: ClientRequest) throws -> ClientAction {
+        return try buffer.parseUnwindingIfNeeded { buffer -> ClientAction in
+            guard let response = try buffer.readServerResponse() else {
+                return .waitForMoreData
+            }
+            guard response.reply == .succeeded else {
+                throw SOCKSError.ConnectionFailed(reply: response.reply)
+            }
+            self.state = .active
+            return .proxyEstablished
+        }
+    }
+    
+    mutating func authenticate(_ buffer: inout ByteBuffer) -> ClientAction {
+        // we don't currently support any authentication
+        // so assume all is fine, and instruct the client
+        // to send the request
+        self.state = .waitingForClientRequest
```

Comment:
This function should probably actually be passed the selected authentication method. For now it can just `precondition` that it's the only value we support. That'll nicely cover the comment we have here.